server/util/qps/qps_test.go

```diff
@@ -0,0 +1,71 @@
+package qps_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/qps"
+	"github.com/stretchr/testify/require"
+)
+
+func TestQPS(t *testing.T) {
+	ticker := make(chan time.Time, 1)
+	counter := qps.NewCounterForTesting(5*time.Second, ticker)
+
+	// Test cases where the ring buffer is not full.
+	require.Equal(t, float64(0), counter.Get())
+	counter.Inc()
+	require.Equal(t, float64(12), counter.Get())
+	for i := 0; i < 59; i++ {
+		counter.Inc()
+	}
+	require.Equal(t, float64(720), counter.Get())
+	for i := 0; i < 29; i++ {
+		ticker <- time.Now()
+	}
+	// The current bin is incremented asynchronously upon receiving a tick via
+	// the ticker channel. Give it some time to run.
+	time.Sleep(10 * time.Millisecond)
```

Comment:
This will likely be flaky. Could do one of these instead
- Have the QPS counter accept an optional `updates` channel (`chan struct{}`) and send a non-blocking message on that channel whenever a tick is processed. Then the test can receive from the channel here instead of waiting for an update.
- Do whitebox testing - place qps in the same package as qps_test, and manually call some `update(now time.Time)` function in the QPS counter rather than driving the counter using a ticker.
- Do a poll here and wait for the condition to become true - the downside though is that the test will timeout if there's a bug rather than immediately failing which is a little annoying.